robot/impl/local_robot_test.go

```diff
@@ -3134,19 +3182,468 @@ func TestMachineStatus(t *testing.T) {
+		name          string
+		localCloudMd  bool
+		remoteCloudMd bool
+	}{
+		{
+			name:          "local cloud metadata and remote cloud metadata",
+			localCloudMd:  true,
+			remoteCloudMd: true,
+		},
+		{
+			name:          "local cloud metadata and no remote cloud metadata",
+			localCloudMd:  true,
+			remoteCloudMd: false,
+		},
+		{
+			name:          "no local cloud metadata and remote cloud metadata",
+			localCloudMd:  false,
+			remoteCloudMd: true,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+
+			resName1 := resource.NewName(resource.APINamespace("acme").WithComponentType("huwat"), "thing1")
+			injectRemoteRobot := &inject.Robot{
+				LoggerFunc:          func() logging.Logger { return logger },
+				ResourceRPCAPIsFunc: func() []resource.RPCAPI { return nil },
+				ResourceNamesFunc:   func() []resource.Name { return []resource.Name{resName1} },
+			}
+			remoteName := "remote1"
+			injectRemoteRobot.ResourceByNameFunc = func(name resource.Name) (resource.Resource, error) {
+				for _, rName := range injectRemoteRobot.ResourceNames() {
+					if rName == name {
+						return rgrpc.NewForeignResource(rName, nil), nil
+					}
+				}
+				return nil, resource.NewNotFoundError(name)
+			}
+
+			remoteMd := cloud.Metadata{
+				PrimaryOrgID:  "the-remote-org",
+				LocationID:    "the-remote-location",
+				MachineID:     "the-remote-machine",
+				MachinePartID: "the-remote-part",
+			}
+			injectRemoteRobot.CloudMetadataFunc = func(context.Context) (cloud.Metadata, error) {
+				if !tc.remoteCloudMd {
+					return cloud.Metadata{}, errNoCloudMetadata
+				}
+				return remoteMd, nil
+			}
+			injectRemoteRobot.MachineStatusFunc = func(ctx context.Context) (robot.MachineStatus, error) {
+				// check that a timeout is passed down from the caller.
+				if _, ok := ctx.Deadline(); !ok {
+					return robot.MachineStatus{}, errors.New("no timeout detected")
+				}
+				md := cloud.Metadata{}
+				if tc.remoteCloudMd {
+					md = remoteMd
+				}
+				return robot.MachineStatus{
+					Resources: []resource.Status{
+						{
+							NodeStatus: resource.NodeStatus{
+								Name: resName1, State: resource.NodeStateUnconfigured, Revision: "rev2",
+							},
+							CloudMetadata: md,
+						},
+					},
+				}, nil
+			}
+			dRobot := newDummyRobot(t, injectRemoteRobot)
+			md := cloud.Metadata{
+				PrimaryOrgID:  "the-org",
+				LocationID:    "the-location",
+				MachineID:     "the-machine",
+				MachinePartID: "the-robot-part",
+			}
+			rev1 := "rev1"
+			cfg := &config.Config{}
+			if tc.localCloudMd {
+				cfg = &config.Config{
+					Cloud: &config.Cloud{
+						ID:           md.MachinePartID,
+						LocationID:   md.LocationID,
+						PrimaryOrgID: md.PrimaryOrgID,
+						MachineID:    md.MachineID,
+					},
+					Revision: rev1,
+				}
+			}
+			lr := setupLocalRobot(t, ctx, cfg, logger, withDisableBackgroundReconfiguration())
+			lr.(*localRobot).manager.addRemote(
```

Comment:
Why `addRemote` instead of putting the remote in the `config.Config` object? Also, can we guard these casts to `*localRobot` so there's never a panic?

Reply:
we're not spinning up new servers, so adding the remote in the config wouldn't actually work